File: android/app/src/main/java/com/apprch/app/data/EventStore.kt
package com.apprch.app.data

import com.apprch.app.ApprchApplication
import com.apprch.app.model.startOfDay
import com.apprch.app.widget.WidgetSnapshotSync
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import java.util.Date

object EventStore {
    suspend fun log(taskId: String, groupId: String) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val taskRef = db.collection("tasks").document(taskId)
        val snapshot = taskRef.get().await()
        if (!snapshot.exists() || snapshot.getString("groupId") != groupId) {
            throw IllegalStateException("This task isn’t available.")
        }

        val batch = db.batch()
        batch.set(
            db.collection("events").document(),
            mapOf(
                "groupId" to groupId,
                "taskId" to taskId,
                "loggedByUid" to uid,
                "timestamp" to FieldValue.serverTimestamp()
            )
        )
        batch.update(
            taskRef,
            mapOf(
                "lastLoggedAt" to FieldValue.serverTimestamp(),
                "lastLoggedByUid" to uid,
                "eventCount" to FieldValue.increment(1)
            )
        )
        batch.commit().await()
        WidgetSnapshotSync.refreshTask(ApprchApplication.appContext, taskId, groupId)
    }

    suspend fun toggleToday(taskId: String, groupId: String, currentlyComplete: Boolean) {
        if (currentlyComplete) undoToday(taskId, groupId) else log(taskId, groupId)
    }

    suspend fun undoToday(taskId: String, groupId: String) {
        FirebaseAuth.getInstance().currentUser
            ?: throw IllegalStateException("Please sign in again.")
        val db = FirebaseFirestore.getInstance()
        val taskRef = db.collection("tasks").document(taskId)
        val snapshot = taskRef.get().await()
        if (!snapshot.exists() || snapshot.getString("groupId") != groupId) {
            throw IllegalStateException("This task isn’t available.")
        }

        val events = db.collection("events")
            .whereEqualTo("groupId", groupId)
            .whereEqualTo("taskId", taskId)
            .get()
            .await()

        val start = startOfDay()
        val today = mutableListOf<com.google.firebase.firestore.DocumentSnapshot>()
        val earlier = mutableListOf<com.google.firebase.firestore.DocumentSnapshot>()
        for (doc in events.documents) {
            val timestamp = doc.getTimestamp("timestamp")
            if (timestamp == null || !timestamp.toDate().before(start)) {
                today += doc
            } else {
                earlier += doc
            }
        }

        val latestEarlier = earlier.maxByOrNull {
            (it.getTimestamp("timestamp")?.toDate() ?: Date(0)).time
        }

        val batch = db.batch()
        today.forEach { batch.delete(it.reference) }

        val currentCount = (snapshot.getLong("eventCount") ?: events.size().toLong()).toInt()
        val update = mutableMapOf<String, Any>(
            "eventCount" to maxOf(0, currentCount - today.size)
        )
        if (latestEarlier != null) {
            latestEarlier.getTimestamp("timestamp")?.let { update["lastLoggedAt"] = it }
            (latestEarlier.getString("loggedByUid") ?: latestEarlier.getString("triggeredByUid"))
                ?.let { update["lastLoggedByUid"] = it }
        } else {
            update["lastLoggedAt"] = FieldValue.delete()
            update["lastLoggedByUid"] = FieldValue.delete()
        }
        batch.update(taskRef, update)
        batch.commit().await()
        WidgetSnapshotSync.refreshTask(ApprchApplication.appContext, taskId, groupId)
    }

    fun userFacingMessage(error: Throwable): String {
        val text = error.localizedMessage.orEmpty()
        if (text.contains("permission", ignoreCase = true) ||
            text.contains("unauthenticated", ignoreCase = true)
        ) {
            return "Couldn’t log this. Check that you’re signed in and try again."
        }
        return text.ifBlank { "Couldn’t log this. Try again." }
    }
}

File: ios/Apprch/ViewModels/EventStore.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

enum EventStore {
    static func log(triggerId: String, groupId: String) async throws {
        guard let uid = Auth.auth().currentUser?.uid else {
            throw EventStoreError.signedOut
        }

        let db = Firestore.firestore()
        let taskRef = db.collection("tasks").document(triggerId)
        let snapshot = try await taskRef.getDocument()
        guard snapshot.exists, snapshot.data()?["groupId"] as? String == groupId else {
            throw EventStoreError.unavailable
        }

        let eventRef = db.collection("events").document()
        let batch = db.batch()
        batch.setData(
            [
                "groupId": groupId,
                "taskId": triggerId,
                "loggedByUid": uid,
                "timestamp": FieldValue.serverTimestamp()
            ],
            forDocument: eventRef
        )
        batch.updateData(
            [
                "lastLoggedAt": FieldValue.serverTimestamp(),
                "lastLoggedByUid": uid,
                "eventCount": FieldValue.increment(Int64(1))
            ],
            forDocument: taskRef
        )
        try await batch.commit()
        await WidgetSnapshotSync.refreshTrigger(id: triggerId, groupId: groupId)
    }

    static func toggleToday(triggerId: String, groupId: String, currentlyComplete: Bool) async throws {
        if currentlyComplete {
            try await undoToday(triggerId: triggerId, groupId: groupId)
        } else {
            try await log(triggerId: triggerId, groupId: groupId)
        }
    }

    static func undoToday(triggerId: String, groupId: String) async throws {
        guard Auth.auth().currentUser?.uid != nil else {
            throw EventStoreError.signedOut
        }

        let db = Firestore.firestore()
        let taskRef = db.collection("tasks").document(triggerId)
        let snapshot = try await taskRef.getDocument()
        guard snapshot.exists, snapshot.data()?["groupId"] as? String == groupId else {
            throw EventStoreError.unavailable
        }

        let eventsSnap = try await db.collection("events")
            .whereField("groupId", isEqualTo: groupId)
            .whereField("taskId", isEqualTo: triggerId)
            .getDocuments()

        let startOfDay = Calendar.current.startOfDay(for: Date())
        var today: [QueryDocumentSnapshot] = []
        var earlier: [QueryDocumentSnapshot] = []
        for doc in eventsSnap.documents {
            guard let timestamp = doc.data()["timestamp"] as? Timestamp else {
                today.append(doc)
                continue
            }
            if timestamp.dateValue() >= startOfDay {
                today.append(doc)
            } else {
                earlier.append(doc)
            }
        }

        let latestEarlier = earlier.max { lhs, rhs in
            let l = (lhs.data()["timestamp"] as? Timestamp)?.dateValue() ?? .distantPast
            let r = (rhs.data()["timestamp"] as? Timestamp)?.dateValue() ?? .distantPast
            return l < r
        }

        let batch = db.batch()
        for doc in today {
            batch.deleteDocument(doc.reference)
        }

        let currentCount = (snapshot.data()?["eventCount"] as? Int)
            ?? (snapshot.data()?["eventCount"] as? Int64).map(Int.init)
            ?? eventsSnap.documents.count
        var taskUpdate: [String: Any] = [
            "eventCount": max(0, currentCount - today.count)
        ]
        if let latestEarlier {
            taskUpdate["lastLoggedAt"] = latestEarlier.data()["timestamp"] as Any
            taskUpdate["lastLoggedByUid"] = (latestEarlier.data()["loggedByUid"] ?? latestEarlier.data()["triggeredByUid"]) as Any
        } else {
            taskUpdate["lastLoggedAt"] = FieldValue.delete()
            taskUpdate["lastLoggedByUid"] = FieldValue.delete()
        }
        batch.updateData(taskUpdate, forDocument: taskRef)
        try await batch.commit()
        await WidgetSnapshotSync.refreshTrigger(id: triggerId, groupId: groupId)
    }

    static func userFacingMessage(for error: Error) -> String {
        if let storeError = error as? EventStoreError {
            return storeError.localizedDescription
        }
        let text = error.localizedDescription
        if text.localizedCaseInsensitiveContains("permission")
            || text.localizedCaseInsensitiveContains("unauthenticated") {
            return "Couldn’t log this. Check that you’re signed in and try again."
        }
        return text
    }
}

enum EventStoreError: LocalizedError {
    case signedOut
    case unavailable

    var errorDescription: String? {
        switch self {
        case .signedOut: return "Please sign in again."
        case .unavailable: return "This task isn’t available."
        }
    }
}
